Parse prm files in python script

gassmoeller opened the pull request:
This is the python script I and some others put together in ASPECT to read in and (optionally) modify .prm files. In its current state it does:

- read in .prm files and store them internally in a tree-like dictionary structure (this makes it possible to move parameters or whole subsections around, I did not include the code to do that here yet)
- reformat the .prm files according to the following rules:
  - 2 spaces for indentation
  - parameters before subsections
  - empty lines before comments
  - combine contents of subsections (no duplicate subsections)
  - later parameters overwrite earlier parameters (no duplicate parameters)
  - retain as much user formatting as possible

To show its impact I ran it on all .prm files included in the deal.II repository. I am not sure if some of those .prm files have certain structures on purpose, so we do not have to apply it on all files. But I think the script is quite useful in updating large numbers of input files and also to enforce a certain similar style across all .prm files in a project.

Happy to modify the PR or discuss how this script could be useful or applied (e.g. during CI?)

gassmoeller commented:
@blaisb @tjhei  you were interested in this, happy to hear what you think.

gassmoeller commented:
As discussed with @blaisb: We should decide if we always want parameters to precede subsections. I personally prefer this optically (and mentally it is easier to first see all parameters of a subsection, before switching context to a subsubsection). However moving parameters before subsections introduces more changes to user input prms than just keeping whatever order the .prm contained. I am open for suggestions.

gassmoeller commented:
It would be great  to get some feedback about people's feelings about this PR and the proposed .prm file format. Of course this is something we may want to think about a bit, but it would help me to know if there is general interest and/or principal opposition to some parts of the new format. The changes to step-35 probably exemplify what could happen to user input files (it all still works, it is just a different layout). If the new layout is not wanted I can also modify the script to keep the order of files intact (even if it is not what I would prefer).

Rombur commented:
I like the idea of having a kind of `clang-format` for the `prm` files in the tutorials. I am in favor of this PR though I don't have strong feelings about the exact format.

blaisb commented:
I am also in favor of this PR. I think a python script is maybe easier to use than an Awk Script. Personnally I merge this as-is

blaisb commented:
> > However, I am not sure if we should modify every existing parameter files with this because if we regenerate the parameter files using the step for examples, we will have to indent them again.
> 
> Wait do we use the parameter handler to generate some of these files? Which ones? And yes, I completely agree, the parameter handler should output the files in the format that the indent script recognizes as "correct" so if there is a mismatch I should adjust the indent script (or the parameter handler).
> 
> I would be in favour of applying the indent script to as many files as possible to make sure they have a uniform format (except of course tests that on purpose want to test a special .prm file, I accidentally updated one of those and reverted the change in the last commit).

Yes, some files are generated by the parameter handler (For example the one for step-70), but remain static in the repository.
Consequently, right now our parameter files do not adhere to the python script indentation standard. It's really not a big deal, but that's something to keep in mind IMO before we indent all of our prm files.

gassmoeller commented:
I updated the PR with the following changes:
- I modified ParameterHandler to not write trailing whitespace in comment lines that are empty (previously `# `, now `#`) and parameters without value (previously `set ... = `, now `set ... =`)
- I modified the script to not modify the order of parameters and subsections (keeping the user order, even if parameters follow subsections)
- I only applied the script to the examples parameter files, and selectively removed some changes that would be different from what the ParameterHandler would write

I think in the current form this is minimally invasive. As a follow-up we should think about how to advertise and/or use the indent script, because I doubt someone will find the script by accident.

blaisb reviewed (approved):
I really like this. I think this is very good work. I woud merge as is.

blaisb commented:
@bangerth and @tjhei are you ok with this version? I think this is a very nice way to move forward.

bangerth commented on contrib/utilities/prm_tools.py:
```diff
@@ -0,0 +1,315 @@
+def split_parameter_line(line):
+    """ Read a 'set parameter' line and extract the name, value, and format. """
+
+    equal_index = line.find('=')
+    # Determine number of additional spaces left of equal sign.
+    # We need to store these separately, because there might be two
+    # lines setting the same parameter, but different number of spaces. They
+    # should still map to the same parameter.
+    alignment_spaces = len(line[:equal_index]) - len(line[:equal_index].rstrip())
+
+    # skip initial word "set" in parameter name
+    words_left_of_equal_sign = line[:equal_index].replace("\\\n","").split()
+    param_name = words_left_of_equal_sign[1:]
+    param_name = ' '.join(param_name)
+
+    # strip spaces at end of value string, keep all inline comments
+    param_value = line[equal_index+1:].rstrip()
+    # if there is one or more spaces at the start of the string, remove one space.
+    # keep additional spaces, because they are part of the formatting.
+    # this way we can add one space when writing the file back and keep the formatting
+    if param_value.startswith(' ') and len(param_value) > 1:
+        param_value = param_value[1:]
+
+    return param_name, param_value, alignment_spaces
```
It perhaps does not matter very much, but I think this all would have been easier if you had just applied a regex of the form `^ *set (.*?)( *)= *(.*?) *$` and then captured the three marked parts. You might have to deal with spaces actually being tabs as well.

bangerth commented on contrib/utilities/prm_tools.py:
```diff
@@ -0,0 +1,315 @@
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+                    prog='deal.II .prm file reformatter',
+                    description='Reformats deal.II .prm files to follow our general formatting guidelines. See the documentation of this script for details.')
```
I think it would be nice to print the intended use (i.e., the list of arguments) here.

bangerth commented on source/base/parameter_handler.cc:
```diff
@@ -1506,15 +1506,23 @@ ParameterHandler::recursively_print_parameters(
+                    out << std::setw(overall_indent_level * 2) << "#";
+
+                    if (!doc_line.empty())
+                      out << " " << doc_line;
```
```suggestion
                      out << ' ' << doc_line;
```

bangerth commented on source/base/parameter_handler.cc:
```diff
@@ -1506,15 +1506,23 @@ ParameterHandler::recursively_print_parameters(
-                  out << std::setw(overall_indent_level * 2) << ""
-                      << "# " << doc_line << '\n';
+                  {
+                    out << std::setw(overall_indent_level * 2) << "#";
```
```suggestion
                    out << std::setw(overall_indent_level * 2) << '#';
```

bangerth commented on source/base/parameter_handler.cc:
```diff
@@ -1506,15 +1506,23 @@ ParameterHandler::recursively_print_parameters(
-                << "= " << value;
+                << "=";
+            if (!value.empty())
+              out << " " << value;
```
```suggestion
              out << ' ' << value;
```

bangerth reviewed (commented):
Everything other than the script is pretty obviously correct and we could merge that separately.

For the script itself, here is a question: If a script contains an `include` statement, what do you do? You can't move things from above to below that statement, nor the other way around.

tjhei commented:
> @bangerth and @tjhei are you ok with this version?

Yes, I think this is a good idea.

gassmoeller commented:
I rebased this PR on top of #17616 to separate the changes to the parameter handler from the python script.

@bangerth to your comments:

> If a script contains an include statement, what do you do? You can't move things from above to below that statement, nor the other way around.

I do not longer change the order of statements. Every include, parameter, and subsection will appear in the output in the same order it was read in.

> It perhaps does not matter very much, but I think this all would have been easier if you had just applied a regex of the form ^ *set (.*?)( *)= *(.*?) *$ and then captured the three marked parts. You might have to deal with spaces actually being tabs as well.

It is possible a regex would have been simpler, but I do not have the time to build one at the moment and there are a lot of special cases to consider and test, e.g. what about parameter values that follow a = sign without whitespace? or parameter values that only consist of whitespace? Parameter names or values that contain line breaks? I spent few days hunting all special cases for the current implementation. I am of course open for improvements to the script or if someone wants to build and test a regex we can certainly use that.

Regarding tabs: I now modified the script to replace every tab with 2 spaces (every tab at the beginning or end of a line gets stripped anyway, so it only concerns tabs around the equal sign -- do we allow tabs in parameter or subsection names?). 2 spaces is our usual indentation in .prm files anyway, and I prefer that while we support tabs, we do not encourage them.

> I think it would be nice to print the intended use (i.e., the list of arguments) here.

I extended the help message to include the intended use and also the formatting guidelines that are applied.

blaisb commented:
ping @bangerth I would like to get this merge soon since @gassmoeller put in some hard work on this :)

bangerth reviewed (approved):
Thanks for making the changes! I had to look up whether the Python `dict` type is sorted or preserves the order to convince myself that you end up writing in the same order as you read -- the answer is that you indeed do (which is of course why you chose a dictionary).

bangerth commented:
The failing test is the usual `trilinos/solver_control_06.mpirun=2.debug`, unrelated to this patch.
